commit c6f5e9955adbc45688074ab1b8cac5d340262c56 (refs/changes/71/3771/5)
Author: John Levon <john.levon@joyent.com>
Date:   2018-04-05T16:50:36+00:00 (1 year, 6 months ago)
    
    OS-6727 bhyve should provide zone UUID to guest via SMBIOS
    Reviewed by: Mike Gerdts <mike.gerdts@joyent.com>
    Approved by: Mike Gerdts <mike.gerdts@joyent.com>

diff --git a/src/sysinfo b/src/sysinfo
index 67873b78..c28215f2 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -103,14 +103,24 @@ function get_smbios_system_info()
 {
     # This puts the variables we're pulling out into the local environment
     eval $(smbios -t SMB_TYPE_SYSTEM | tr -d "\"" \
-        | egrep "Manufacturer: |Product: |Serial Number: |UUID: |SKU Number: |Version: |Family: " \
+        | egrep "Manufacturer: |Product: |Serial Number: |UUID \(Endian-corrected\): |UUID: |SKU Number: |Version: |Family: " \
         | sed -e 's/^ *//' \
         -e 's/: /="/' \
         -e 's/ *$/"/' \
         -e 's/Serial Number/Serial_Number/' \
         -e 's/SKU Number/SKU_Number/' \
         -e 's/Version/HW_Version/' \
-        -e 's/Family/HW_Family/')
+        -e 's/Family/HW_Family/' \
+        -e 's/UUID (Endian-corrected)/Fixed_UUID/')
+
+    vendor=$(smbios -t SMB_TYPE_BIOS | grep "Vendor: " \
+        | sed -e 's/^.*Vendor: //;s/ *$//')
+
+    # If we are a bhyve guest, we can use the correct UUID format; otherwise we
+    # use the network-endian one for compatibility reasons.
+    if [ "$vendor" = "BHYVE" -a -n "$Fixed_UUID" ]; then
+        UUID=$Fixed_UUID
+    fi
 
     # overwrite UUID if config dictates otherwise
     tmp_uuid=$(/usr/bin/bootparams | grep "^override_uuid=" | cut -f2 -d'=')
